test/filter.test.ts

```diff
@@ -0,0 +1,94 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+import { expectAssignable, expectNotType, expectType } from 'tsd';
+import { __, compose, filter, identity, isNotNil, pipe, map } from '../es';
+
+type StringOrUndefined = string | undefined;
+
+// when the predicate that is passed to filter has a generic argument, such as `isNotNil`
+// typescript collapses `T` to `unknown`
+expectAssignable<unknown[]>(filter(isNotNil)([] as StringOrUndefined[]));
+expectAssignable<unknown[]>(filter(isNotNil, [] as StringOrUndefined[]));
```

Comment:
Why does `T` collapse to `unknown`?

I would expect these to be `string[]`.  Can `string[]` be inferred somehow?

Reply:
I don't know. But I can simulate when it does and does not

See this playground: https://tsplay.dev/NdaA0m